src/stream/src/executor/hash_join.rs

```diff
@@ -821,246 +839,389 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
-                                }
-                            } else {
-                                for (column_idx, watermark) in &useful_state_clean_columns {
-                                    if matched_row.row.datum_at(*column_idx).map_or(
-                                        false,
-                                        |scalar| {
-                                            scalar
-                                                .default_cmp(&watermark.val.as_scalar_ref_impl())
-                                                .is_lt()
-                                        },
-                                    ) {
-                                        need_state_clean = true;
-                                        break;
-                                    }
-                                }
-                            }
-                            // If the stream is append-only and the join key covers pk in both side,
-                            // then we can remove matched rows since pk is unique and will not be
-                            // inserted again
-                            if append_only_optimize {
-                                // Since join key contains pk and pk is unique, there should be only
-                                // one row if matched.
-                                assert!(append_only_matched_row.is_none());
-                                append_only_matched_row = Some(matched_row);
-                            } else if need_state_clean {
-                                // `append_only_optimize` and `need_state_clean` won't both be true.
-                                // 'else' here is only to suppress compiler error.
-                                matched_rows_to_clean.push(matched_row);
-                            }
-                        }
-                        if degree == 0 {
-                            if let Some(chunk) =
-                                hashjoin_chunk_builder.forward_if_not_matched(Op::Insert, row)
-                            {
-                                yield chunk;
-                            }
-                        } else if let Some(chunk) =
-                            hashjoin_chunk_builder.forward_exactly_once_if_matched(Op::Insert, row)
-                        {
-                            yield chunk;
-                        }
-                        // Insert back the state taken from ht.
-                        side_match.ht.update_state(key, matched_rows);
-                        for matched_row in matched_rows_to_clean {
-                            if side_match.need_degree_table {
-                                side_match.ht.delete(key, matched_row)?;
-                            } else {
-                                side_match.ht.delete_row(key, matched_row.row)?;
-                            }
-                        }
```

Comment:
This old section of code is entirely removed, because it contains a lot of duplicated logic between the `Insert / Delete` arms. That made it hard to modify.

Hence we refactor it into some common functions and macros.